keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+notes right away. Also the kubeadm maintainers might make mistakes forgetting
+about a certain feature/bug fix.
+
+### Test Plan
+
+#### CI artifacts
+
+Currently the kubeadm binary is built and uploaded in the CI GCS bucket
+next to the rest of the k8sbinaries like kubelet.
+
+There is a complication as now another repository has to be monitored
+for changes. Currently the CI build happens after a PR merged in the
+k/k repository.
+
+The proposal is to always rebuild kubeadm based on the latest tips of
+the master and release-x.yy branches of the k/kubeadm repository after
+a CI build for the k/k repository triggers. In this case the k/release
+build tooling must not trigger the writing of a KUBE_VERSION file in the
+k/kubeadm repository but rather let the version to generate based on the
+latest Git tags and SHAs.
+
+#### PR testing
+
+There is already support for running pre-submit jobs on k/kubeadm PRs.
+Once kubeadm is moved to the new location a collection of PR blocking
+jobs should be added to make sure the project is always buildable with
+the current version of Go that the k/k repository uses. Syncing Go versions
+should be the responsibility of the k/kubeadm maintainers.
+
+The new PR blocking jobs must support building kubeadm from source with
+the latest commits that a PR for k/kubeadm introduces. The rest of the
+k8s artifacts can be pulled from the CI GCS bucket. This can also allow
+catching problems introduced in k/k when using kubeadm as the deployer.
+
+The /hack/verify-* and update-* scripts that k/kubeadm/kubeadm uses
+will not be a direct copy of k/k but will provide similar results:
+Go versions are in sync Go-fmt, linter and import tools are in sync
+as much as possible
+
+The following lists of tests will be executed on PRs as blocking:
+- Unit tests
+- Integration tests (currently reside in k/k/cmd/kubeadm/test)
+- E2e tests from the e2e-kubeadm suite (currently resides
+in k/k/test/e2e-kubeadm)
+- Full cluster deployment e2e, possibly upgrade and skew tests too
+
+#### Periodics jobs
+
+No changes to the current kubeadm periodic jobs will be required
+as long as the CI build process continues to push the kubeadm
+and the rest of the k8s binaries to the GCS buckets.
+
+### Graduation Criteria
+
+Not applicable as the work in this KEP should be ideally executed within
+one k8s release cycle.
+
+### Upgrade / Downgrade Strategy
+
+No changes to the upgrade strategy will be required.
+Downgrades are not supported by kubeadm.
+
+### Version Skew Strategy
+
+No changes will be applied to the current version skew support in kubeadm
+as long as kubeadm preserves its exiting versioning, which matches
+the versioning of k8s.
+
+## Implementation History
+
+- 2019-12-29: This proposal document was created
+
+## Open Questions
+
+#### Can we stop building DEBs and RPMs from k/k?
+
+Currently the Kubernetes CI build process uses Bazel specs from the
+k/k `/build` directory to build DEBs and RPMs. Packages are generated
```

Comment:
Suggest adding for completeness following sentence no. 1 a sentence no. 2, eg:

Currently the Kubernetes non-CI build process for official releases uses automation similar to k/k's, but in the k/release repo as of tag `v0.2.1`, and a third variant is in development on k/release repo's master branch.

Reply:
i'm looking for yours and @justaugustus 's response to this question.

Reply:
The answer is yes.  There are three or four candidates for build tooling today.  None of them need to live in k/k.

Reply:
i think the questions is related to the timeline of the KEP too - 1.18
can this happen in parallel in 1.18 if kubeadm starts moving out or are there major blockers that will cause this to shift in the distant future e.g. 1.25?

reminding that SCL are willing to help/staff.

Reply:
I didn't articulate my concern here clearly:

Reply:
i will add more detail about the rest of the build process where release DEBs and RPMs happen in k/release. also will explicitly mention that we can work on that during 1.19.